### PR TITLE
ref: searchbar alert rule condition form

### PR DESCRIPTION
**Pull request by nhsiehgit**
per: https://github.com/getsentry/sentry/issues/75011
Aug-06-2024 14-00-48

### CODE REVIEW

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 1]
```diff
@@ -141,6 +218,36 @@ class RuleConditionsForm extends PureComponent<Props, State> {
     }
   }
 
+  getEventFieldValues = async (tag, query): Promise<string[]> => {
```
Comment (nhsiehgit) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:221]: This is essentially a rewrite of the `searchBar.tsx` method `getEventFieldValues`

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 2]
```diff
@@ -118,6 +146,55 @@ class RuleConditionsForm extends PureComponent<Props, State> {
     this.fetchData();
   }
 
+  getFilterKeys = () => {
```
Comment (nhsiehgit) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:149]: This is essentially a rewrite of the `searchBar.tsx` method `getTagList`

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 3]
```diff
@@ -118,6 +147,55 @@ class RuleConditionsForm extends PureComponent<Props, State> {
...
+          kind: FieldKind.MEASUREMENT,
+        };
+        return measurement_tags;
+      },
+      {}
+    );
+    const orgHasPerformanceView = organization.features.includes('performance-view');
+    const combinedTags: TagCollection =
+      dataset === Dataset.ERRORS
+        ? Object.assign({}, [], STATIC_FIELD_TAGS_WITHOUT_TRANSACTION_FIELDS)
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:166]: i think i'd use spread instead, also i'm not sure the array does anything here
```suggestion
        ? Object.assign({}, STATIC_FIELD_TAGS_WITHOUT_TRANSACTION_FIELDS)
```

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 4]
```diff
@@ -99,15 +121,22 @@ type Props = {
...
     environments: null,
+    measurements: {},
+    filterKeys: {},
   };
 
   componentDidMount() {
     this.fetchData();
+    const measurements = getMeasurements();
+    const filterKeys = this.getFilterKeys();
+    this.setState({measurements, filterKeys});
```
Comment (malwilley) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:139]: Do these need to be in state? Are you doing this to memoize them?

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 5]
```diff
@@ -618,76 +726,103 @@ class RuleConditionsForm extends PureComponent<Props, State> {
...
-                          if (e.key === 'Enter') {
-                            e.preventDefault();
-                            e.stopPropagation();
+                      {organization.features.includes('search-query-builder-alerts') ? (
+                        <SearchQueryBuilder
+                          initialQuery={initialData?.query ?? ''}
+                          getTagValues={this.getEventFieldValues}
+                          placeholder={this.searchPlaceholder}
+                          searchSource="alert_builder"
+                          filterKeys={filterKeys}
```
Comment (malwilley) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:735]: This file has a `getSupportedAndOmittedTags` function that it was using before. Does your new function do the same thing or do we will need that?

File: static/app/views/alerts/rules/metric/ruleForm.tsx [hunk 6]
```diff
@@ -1212,36 +1212,36 @@ class RuleFormContainer extends DeprecatedAsyncComponent<Props, State> {
         >
           <List symbol="colored-numeric">
             <RuleConditionsForm
-              project={project}
+              activationCondition={activationCondition}
```
Comment (nhsiehgit) [inline static/app/views/alerts/rules/metric/ruleForm.tsx:1215]: this is just alphabetizing the props

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 7]
```diff
@@ -618,76 +726,103 @@ class RuleConditionsForm extends PureComponent<Props, State> {
...
+                          {...getSupportedAndOmittedTags(dataset, organization)}
+                          includeSessionTagsValues={dataset === Dataset.SESSIONS}
+                          disabled={disabled || isErrorMigration}
+                          useFormWrapper={false}
+                          organization={organization}
+                          placeholder={this.searchPlaceholder}
+                          onChange={onChange}
+                          query={initialData.query}
+                          // We only need strict validation for Transaction queries, everything else is fine
+                          highlightUnsupportedTags={
```
Comment (malwilley) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:797]: Looks like you probably want to pass this same thing to `disallowUnsupportedFilters` in SearchQueryBuilder?

**Comment by malwilley** [comment]
Something might be wrong with the tag fetching function - the old experience returned some values for `release.build` for example:

CleanShot 2024-08-06 at 16 41 40

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx (hunk 4 above)
Comment (nhsiehgit) [reply static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:139]: Ya pretty much

**Comment by nhsiehgit** [comment]
🤔 
Ah hmm i see, so that's what this function does 👀

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx (hunk 5 above)
Comment (nhsiehgit) [reply static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:735]: Ah got it 👍 
yes, i'll add this to the filterKeys

**Comment by nhsiehgit** [comment]
@malwilley I still see it making the appropriate request to fetch the tags?
Is there any better way to validate this?

**Comment by malwilley** [comment]
We discussed on slack already, but just verifying that the tag value function _does_ work, and actually works much better now that it's sending the project ID 🙌

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx (hunk 4 above)
Comment (malwilley) [reply static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:139]: Okay so I do like that `filterKeys` is memoized, but it looks like this is preventing custom tags from being shown. `this.getFilterKeys` depends on `tags` which are passed in as a prop and those can change depending on what project is displayed. So I think you'll need to make sure that this reacts to changes in `tags`

**Review by malwilley — COMMENTED** [review]
This looks good aside from custom tags not being shown (see screenshots):

CleanShot 2024-08-07 at 10 22 31

CleanShot 2024-08-07 at 10 22 18

The only other thing that could be improved is using `filterKeySections` to organize the keys, but that could be done in a subsequent PR

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 8]
```diff
@@ -118,6 +147,61 @@ class RuleConditionsForm extends PureComponent<Props, State> {
...
+    );
+    const orgHasPerformanceView = organization.features.includes('performance-view');
+    const combinedTags: TagCollection =
+      dataset === Dataset.ERRORS
+        ? Object.assign({}, STATIC_FIELD_TAGS_WITHOUT_TRANSACTION_FIELDS)
+        : dataset === Dataset.TRANSACTIONS
+          ? Object.assign(
+              {},
+              measurementsWithKind,
+              [],
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:171]: i don't think array here does anything

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 9]
```diff
@@ -118,6 +147,61 @@ class RuleConditionsForm extends PureComponent<Props, State> {
...
+              )
+            : Object.assign({}, STATIC_FIELD_TAGS_WITHOUT_TRACING);
+
+    const tagsWithKind = Object.keys(tags).reduce((acc, key) => {
+      acc[key] = {
+        ...tags[key],
+        kind: FieldKind.TAG,
+      };
+      return acc;
+    }, {});
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:191]: types here could be a bit better i think right now the result is `{}`

```suggestion
    const tagsWithKind = Object.keys(tags).reduce<Record<string, something>>((acc, key) => {
      acc[key] = {
        ...tags[key],
        kind: FieldKind.TAG,
      };
      return acc;
    }, {});
```

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 10]
```diff
@@ -118,6 +147,61 @@ class RuleConditionsForm extends PureComponent<Props, State> {
...
+
+    const {omitTags} = {
+      ...getSupportedAndOmittedTags(dataset, organization),
+    };
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:195]: i think spread can be removed
```suggestion
    const {omitTags} = getSupportedAndOmittedTags(dataset, organization);
```

**Review by scttcper — APPROVED** [review]

**Pull request merged**